Integrate `zephyr-xenlib` to Zephyr's main repository

soburi opened the pull request:
# Summary

Proposing adoption of this project as a standard Zephyr module, to be merged into the main Zephyr repository.
Once approved and merged, I plan to apply for the repository to be added to Zephyr.

This change mainly imports Xen’s public headers.

Related Zephyr side work is tracked in https://github.com/zephyrproject-rtos/zephyr/pull/91643.

# Details

Zephyr recently tightened its policies against including non-Apache-2.0 licensed files in the main tree, as documented at https://github.com/zephyrproject-rtos/zephyr/tree/main/LICENSES

Adding any new license now requires TSC approval and is enforced by CI. Adding the MIT-licensed Xen public headers directly into the main repository is becoming cumbersome. Modules outside the main tree are exempt from this restriction, so the headers need to live in an external module instead.

The `zephyr-xenlib` module already brings in Xen public headers separately for the `vch` and `xenstore-srv` components. This proposal generalizes that approach so the module itself carries the Xen public headers, and it updates include paths accordingly.

# Verification

Functionally, this is purely a relocation of headers: no APIs or implementations change. However, when importing the Xen public headers, I applied local adjustments; please focus review attention there. (https://github.com/xen-troops/zephyr-xenlib/pull/102/commits/ea9b3eb71e0a35cbc8bd2d0ada2d5a58db4825e0)

I've also reapplied the commit history in the Zephyr repository to ensure a clear history. However, since it eventually replaces the content with that of xen-4.17, I'll squash if it was redundant.

The module has been validated in the https://github.com/xen-troops/meta-xt-prod-devel-rpi5 environment, building and working fine without modifications.

# Response on the Zephyr side

This module is consumed by https://github.com/xen-troops/meta-xt-prod-devel-rpi5
, which derives from Zephyr 3.6.0 and receives feature updates on the branch https://github.com/xen-troops/zephyr/tree/zephyr-v3.6.0-xt
On the Zephyr side, we will cherry-pick this change and continue the upstream integration toward main. https://github.com/zephyrproject-rtos/zephyr/pull/91643)

firscity committed:
xenvm: drivers: xen: add Xen grant table driver

This commit introduces driver for granting access for own grant
table and for mapping/unmapping foreign gref. Grant tables are used
for data exchange between Xen domains via shared memory page(s) (e.g.
for sharing ring buffer with driver data) This functionality is
widely used and needed for implementing PV backend/frontend drivers.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

firscity committed:
xenvm: evtchn: expand Xen event channel driver functionality

This commit adds new functions, which can be used for Xen event channel
management (allocation, interdomain binding etc.). Such functionality
is needed for Xen PV driver development in Zephyr.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

Commit:
include: Prefix includes to use a scope

Move include paths and add new target_include_directories to support
backwards compatibility:
* /include -> /include/zephyr
  example: <irq.h> -> <zephyr/irq.h>

Issue #41543

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Yuval Peress <peress@google.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
include: Import Xen public header

Import Xen public header from RELEASE-4.21.1

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Mykola Kvach <mykola_kvach@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

gmarull committed:
include: migrate includes to <zephyr/...>

In order to bring consistency in-tree, migrate all includes within
include directory to the new prefix <zephyr/...>. Note that the
conversion has been scripted, refer to zephyrproject-rtos#45388 for more
details.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

firscity committed:
xen: dom0: update Xen public headers for running Zephyr as Dom0

Xen public headers are added to Zephyr kernel partially, so we need
to update them for new features implementation.

Further implementation of Xen domains configuration and memory
management requires dedicated Xen public headers inside Zephyr kernel.
It will be used for add Zephyr OS Xen Domain-0 funtionalities.

Update existent and add new required public headers from Xen 4.17.0
release.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

dsemenets committed:
drivers: xen: add xen version hypercall

Xen API contains hypercall, which allows domains to identify Xen
version, that is currently used on the system. It can be used to check
if current version is supported by Zephyr or to change behavior of the
drivers or services.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

dsemenets committed:
drivers: xen: dom0: add Xen sysctl hypercall

This hypercall can be used get some information about physical machine
and running guests:

- sysctl hypercall "xen_sysctl_getphysinfo" allows read information about
physical machine: number CPUs, memory sizes, hardware capabilities, etc.

- sysctl hypercall "xen_sysctl_getdomaininfolist" returns array of domain
info structures that provide information about particular domain(s).

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

dsemenets committed:
drivers: xen: dom0: added getvcpuinfo domctl call

Domain control call "XEN_DOMCTL_getvcpuinfo" provides information
about domain's virtual CPU(s). It can be used by Domain-0 services
for gathering statistic and vCPUs current status.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

firscity committed:
include: xen: add gnttab_query_size struct to public headers

Xen public headers were imported into Zephyr source tree and only used
structs were added (not copied as is). Mow, for refactoring and
improving gnttab driver we need to augment grant table public header
with one more struct. It will be used for reading actual number of
available Xen grant frames.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

Deedone committed:
xen: Make XEN_INTERFACE_VERSION configurable

Add Kconfig option to specify Xen interface version to use. This will
make it easier to switch between different versions of Xen.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

Deedone committed:
xen: Add support for changing Xen Domctl interface version

Add a new Kconfig option CONFIG_XEN_DOMCTL_INTERFACE_VERSION that allows
to change the version of the Domctl interface used by Zephyr to issue
domctl hypercalls. Add compile-time checks to enable or disable certain
Domctl operations based on the selected Domctl interface version.
For now versions 0x15, 0x16, and 0x17 are supported.

Also it required to correctly guard domctl call that were not supported
prior to specified version.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

Deedone committed:
xen: Add support for changing Xen Sysctl interface version

Add a new Kconfig option CONFIG_XEN_SYSCTL_INTERFACE_VERSION that allows
to change the version of the Sysctl interface used by Zephyr to issue
sysctl hypercalls.
For now versions 0x15 is supported.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

ndrs-pst committed:
include: fix typo in multiple directories

Utilize a code spell-checking tool to scan for and correct spelling errors
in all files within the `include/zephyr` directory.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: xen: add DMOP hypercall wrappers

Add wrappers for following XEN_DMOP_* hypercalls.
These enables Xen device model control path:
dm_op provides operations to create/manage the ioreq server
so guest MMIO accesses are trapped and handled by the hypervisor.
These are guarded by CONFIG_XEN_DMOP.

- dmop
  - dmop_create_ioreq_server
    XEN_DMOP_create_ioreq_server
  - dmop_map_io_range_to_ioreq_server
    XEN_DMOP_map_io_range_to_ioreq_server
  - dmop_set_ioreq_server_state
    XEN_DMOP_set_ioreq_server_state
  - dmop_nr_vcpus
    XEN_DMOP_nr_vcpus
  - dmop_set_irq_level:
    XEN_DMOP_set_irq_level

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

Commit:
xen: arch-arm: sync xen_arch_domainconfig with Xen

The xen_sysctl_getdomaininfo function (which calls the
XEN_SYSCTL_getdomaininfolist hypercall) returned meaningless data
because the xen_domctl_getdomaininfo structure in Zephyr was 8 bits
smaller than in Xen.

This happened because the xen_arch_domainconfig structure, which is used
in both xen_domctl_getdomaininfo and xen_domctl_createdomain, was
missing the uint8_t sve_vl field introduced in Xen Domctl interface
version 0x16.

To fix this, the sve_vl field was added to xen_arch_domainconfig under
the appropriate config macro, synchronizing the structure with Xen and
resolving the data mismatch.

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: Svitlana Drozd <svitlana_drozd@epam.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: xen: memory: add acquire_resource wrapper

Add XENMEM_acquire_resource wrapper to map device model related buffers.
This is required for communication with the ioreq server.

- memory
  - xendom_acquire_resource:
    HYPERVISOR_memory_op(XENMEM_acquire_resource)

Note:
This commit was reapplied when moving from zephyr repository into here.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
public: Migrating public headers used in zephyr-xenlib

Migrate the public headers imported by each modules to shared headers.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@firscity @klogg @harunobukurokawa

Due to my mistake, I deleted my fork repository, which resulted in the #102 being closed, so I recreated it.

soburi commented:
@firscity 

https://github.com/zephyrproject-rtos/zephyr/pull/103152 will probably be included as soon as the 4.4.0 release is complete, so let's resume this discussion.

https://github.com/xen-troops/zephyr-xenlib/pull/103#issuecomment-3612073819

>    Next is regarding Xen public headers: we decide to proceed with your proposal to make zephyr-xenlib a module of zephyrproject-rtos repo. When it will be done in current implementation, we will move all public headers to zephyr-xenlib. I believe it will be done as a "initial" commit with some Xen Release public headers as is, we may discuss which one we should peek.

First, as a starting point for discussion, I've created a simple vanilla version with the files moved.

soburi commented on include/xen/public/domctl.h:
```diff
@@ -0,0 +1,534 @@
+	uint32_t flags;
+
+#define _XEN_DOMCTL_IOMMU_no_sharept	0
+#define XEN_DOMCTL_IOMMU_no_sharep	(1U << _XEN_DOMCTL_IOMMU_no_sharept)
```
For now, the policy is to leave things as they are.

soburi commented on include/xen/public/io/console.h:
```diff
@@ -0,0 +1,48 @@
+
+#ifndef __XEN_PUBLIC_IO_CONSOLE_H__
+#define __XEN_PUBLIC_IO_CONSOLE_H__
+
```
For now, the policy is to leave things as they are.

soburi commented on include/xen/public/hvm/ioreq.h:
```diff
@@ -0,0 +1,125 @@
+	uint32_t count;    /* for rep prefixes */
+	uint32_t size;     /* size in bytes */
+	uint32_t vp_eport; /* evtchn for notifications to/from device model */
+	uint16_t _pad0;
```
For now, the policy is to leave things as they are.

soburi commented on include/xen/public/io/console.h:
```diff
@@ -0,0 +1,48 @@
+	XENCONS_RING_IDX out_cons, out_prod;
+};
+
+#ifdef XEN_WANT_FLEX_CONSOLE_RING
+#include "ring.h"
+DEFINE_XEN_FLEX_RING(xencons);
+#endif
+
```
For now, the policy is to leave things as they are.

firscity commented:
Hi @soburi, thank you for your work and sorry for late reply!

Regarding public header migration - per my understanding, we have 2 final goals:
1. Make zephyr-xenlib a part of zephyrproject-rtos;
2. Move public header to zephyr-xenlib (and remove from Zephyr RTOS include/)

Also, it may be connected with 2 other issues:
- xen drivers are located inside Zephyr RTOS repo and I don't now if it is correct to use external headers from internal driver
- public headers is from Xen RELEASE-4.19.0 and it is good time to update it during transfer

Please, let me know your view on following points and how you see further steps.

CC: @drozd-svtln

soburi commented:
> Hi @soburi, thank you for your work and sorry for late reply!
> 
> Regarding public header migration - per my understanding, we have 2 final goals:
> 
>     1. Make zephyr-xenlib a part of zephyrproject-rtos;
> 
>     2. Move public header to zephyr-xenlib (and remove from Zephyr RTOS include/)
> 
> 
> Also, it may be connected with 2 other issues:
> 
>     * xen drivers are located inside Zephyr RTOS repo and I don't now if it is correct to use external headers from internal driver
> 
>     * public headers is from Xen RELEASE-4.19.0 and it is good time to update it during transfer
> 
> 
> Please, let me know your view on following points and how you see further steps.
> 
> CC: @drozd-svtln

>     * xen drivers are located inside Zephyr RTOS repo and I don't now if it is correct to use external headers from internal driver

I believe this is not a problem.
That is, hal is naturally used by the drivers, and picolibc is referenced by all modules.


>     * public headers is from Xen RELEASE-4.19.0 and it is good time to update it during transfer

I agree with this as well. However, I think the migration itself should be divided into two phases. That is, the first phase should be simple files moving and integrating modules (This PR), followed by the second phase of updating the headers.
It would be troublesome if changes were made together during module integration.

The introduction of the Xen module was previously proposed and approved. Re-issuing this proposal should minimize confusion. https://github.com/zephyrproject-rtos/zephyr/issues/91644

Fortunately or unfortunately, we won't be able to move forward until April 16th, when the 4.4.0 release is complete.
Let's think about it during that time.

firscity commented:
> I believe this is not a problem.
That is, hal is naturally used by the drivers, and picolibc is referenced by all modules.

Oh, I see, I'll check it and maybe try on my side. I will be happy to find a sequence of actions for seamless migration (like integrating zephyr-xenlib to main repo -> change includes to zephyr-xenlib -> remove public headers from Zephyr includes or smth).

> I agree with this as well. However, I think the migration itself should be divided into two phases. That is, the first phase should be simple files moving and integrating modules (This PR), followed by the second phase of updating the headers.
It would be troublesome if changes were made together during module integration.
The introduction of the Xen module was previously proposed and approved. Re-issuing this proposal should minimize confusion. https://github.com/zephyrproject-rtos/zephyr/issues/91644

Yeah, I remembers about TSC approval, but it was only regarding Xen headers, not about the whole zephyr-xenlib and, as a result, new repo on zephyrproject-rtos. Will it require any separate approval? Unfortunately, I'm a bit distant from Zephyr meetings and current practicies...

> Fortunately or unfortunately, we won't be able to move forward until April 16th, when the 4.4.0 release is complete.
Let's think about it during that time.

I believe fortunately, because due to high load on other work activities I can't spent much time on this, but hope I will finally have time to review and integrate Xenstore-client interface PR...

Thank you for your work and help!

soburi commented:
@firscity 

Sorry for the late response, I've been a bit busy.

> Oh, I see, I'll check it and maybe try on my side. I will be happy to find a sequence of actions for seamless migration (like integrating zephyr-xenlib to main repo -> change includes to zephyr-xenlib -> remove public headers from Zephyr includes or smth).

Regarding this, IMO,
1. Create a branch for migrating zephyr-xenlib.
2. Request to create a fork on the Zephyr side based on that branch. (Updating Issue https://github.com/zephyrproject-rtos/zephyr/issues/91644)
3. Update zephyr's west.yml to reference the created fork.

So, I would like you to create a branch from the current main for the purpose of 1. I will resubmit this PR for new one.

> Yeah, I remembers about TSC approval, but it was only regarding Xen headers, not about the whole zephyr-xenlib and, as a result, new repo on zephyrproject-rtos. Will it require any separate approval? Unfortunately, I'm a bit distant from Zephyr meetings and current practicies...

Of course, further explanation will be necessary, but reporting the result of scancode should suffice.

firscity commented:
Hi @soburi,

Sorry for delays in replays, I checked both this and #103 PRs and have some comments. I like the approach with Xen public headers in zephyr-xenlib, that will be a part of zephyrproject-rtos repository. It will allow us to use it in both zephyr-xenlib and Zephyr RTOS drivers. But I'm not sure if we should take all of the history for these headers in the way it is done in this PR.

In my opinion, we should create a separate directory in zephyr-xenlib, where we will place Xen public headers as is from certain Xen version (e.g. RELEASE-4.21.0). Then change small parts (like make INTERFACE_VERSION configurable from Zephyr Kconfigs etc) in a separate commit and switch all of the sources to use them from created directory as is. It will be easier to track and update them between Xen Releases, which will be necessary when we will became a part of zephyrproject-rtos I believe. 

Such approach will simplify this PR and also will not require to add new headers in #103.

I want to hear your and probably @lorc thoughts about this here and if everyone agree, this will unblock merging of this and Xenstore related PRs. After this we can start a process to move to another repo.

Thank you for your work and for waiting!

soburi committed:
treewide: Migrating public headers used in zephyr-xenlib

Migrate the public headers imported by each modules to shared headers.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Mykola Kvach <mykola_kvach@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

soburi committed:
include: Adapt xen-compat.h for zephyr

- Define __XEN_TOOLS__ if CONFIG_XEN_DOM0 enabled
- Define __XEN_INTERFACE_VERSION__ from CONFIG_XEN_INTERFACE_VERSION
- Add the common include files needed by the Xen public headers

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Mykola Kvach <mykola_kvach@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

firscity commented on include/xen/public/hvm/hvm_op.h:
```diff
@@ -142,15 +142,15 @@ typedef struct xen_hvm_xentrace xen_hvm_xentrace_t;
 
 /* Following tools-only interfaces may change in future. */
-#if defined(__XEN__) || defined(__XEN_TOOLS__)
+#ifdef CONFIG_XEN_DOM0
```
Probably this is a good time to finally make smth like:
```
#ifdef CONFIG_XEN_DOM0
#define __XEN_TOOLS__
#endif
```
to minimize changes to headers. @lorc what do you think?

soburi replied:
> Probably this is a good time to finally make smth like:
> 
> ```
> #ifdef CONFIG_XEN_DOM0
> #define __XEN_TOOLS__
> #endif
> ```
> 
> to minimize changes to headers. @lorc what do you think?

I'm completely agree it, and I updated it.
The headers keep as original without no modification.
It's quite cool.

firscity reviewed (commented):
Only 1 comment, overall it looks great, thank you for your work!
Lets wait for @lorc before changing anything. After merging this one we will unblock all others, that are blocked by public headers.

firscity commented on CMakeLists.txt:
```diff
@@ -3,6 +3,10 @@
 
+set(ZEPHYR_XENLIB_DIR ${ZEPHYR_CURRENT_MODULE_DIR})
+
+zephyr_include_directories_ifdef(CONFIG_XEN ${ZEPHYR_XENLIB_DIR}/include)
```
Nit: I don't think that `..._ifdef()` is needed here, I think no one will try to build it without `CONFIG_XEN`, so may be unconditional `zephyr_include_directories()`

soburi replied:
When enabled as a Zephyr module, this CMakeLists.txt file is always parsed.
Since there may be cases where builds are performed without Xen enabled in the same environment, it's helpful to leave this condition in place.

However, with Zephyr modules, it's common practice to enclose the entire code in if...endif statements, so I've modified it to follow that convention.

firscity reviewed (approved):
W or wo my comment fixed, please add my `Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>` to all commits.

xakep-amatop commented on include/xen/public/io/libxenvchan.h:
Non-blocking note: the previous local `zephyr-xenlib` copy of `libxenvchan.h` had an include guard, while the newly imported shared copy matches upstream Xen and has no guard.

I do not think we should add a Zephyr-local delta here if the goal is to keep Xen public headers close to upstream. However, this is a small include-hygiene regression compared to the previous local copy, so it may be worth fixing in Xen upstream and then importing the fixed header later.

soburi replied:
Thank you for your review.
It is undesirable there are differences with release version,
so I think it is good to correct in upstream, and incorporate at next update.

xakep-amatop commented on CMakeLists.txt:
```diff
@@ -3,11 +3,19 @@
-add_subdirectory_ifdef(CONFIG_XEN_VCH vch)
-add_subdirectory_ifdef(CONFIG_XEN_LIBFDT libfdt)
-add_subdirectory_ifdef(CONFIG_XSTAT xstat)
+if(CONFIG_XEN)
```
The top-level CMake now makes the whole module conditional on `CONFIG_XEN`.
That is probably fine if all xenlib features are intended to be usable only
when Zephyr's Xen support is enabled.

Should the Kconfig options mirror that by adding `depends on XEN` to the
module features, or by wrapping the xenlib menu in `if XEN`? Otherwise options
such as `XEN_STORE_SRV`, `XEN_DOMAIN_MANAGEMENT`, `XEN_VCH`, or `XSTAT` remain
selectable even though the module sources/include path are ignored when
`CONFIG_XEN=n`.

soburi replied:
It make a sense. I updated it.

And `mainmenu` element not be here, removed it.

lorc reviewed (approved):
Thank you, Tokita-san,


`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`
